File: src/modifiers/contentMargins.ts
import { createModifier } from './createModifier';

/// Edge name(s) the margin applies to. A single edge, one of the grouping
/// keywords, or an array of edge names — parsed natively via the shared
/// edge parser. Defaults to `all`.
export type ContentMarginEdge =
  | 'top'
  | 'bottom'
  | 'leading'
  | 'trailing'
  | 'horizontal'
  | 'vertical'
  | 'all';

export interface ContentMarginsParams {
  /// Edges to inset. Defaults to `all`.
  edges?: ContentMarginEdge | ContentMarginEdge[];
  /// Margin length in points.
  length: number;
}

/// SwiftUI `.contentMargins(_:_:for:)`. Adds margins around the content of
/// a scrollable view (applied for the `.scrollContent` placement). Gated
/// to watchOS 10+ natively; a no-op on older systems.
export function contentMargins(
  length: number
): ReturnType<typeof createModifier>;
export function contentMargins(
  params: ContentMarginsParams
): ReturnType<typeof createModifier>;
export function contentMargins(a: number | ContentMarginsParams) {
  if (typeof a === 'number') {
    return createModifier('contentMargins', { edges: 'all', length: a });
  }
  return createModifier('contentMargins', { edges: 'all', ...a });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/ContentMargins.swift
import SwiftUI

/// SwiftUI `.contentMargins(_:_:for:)`. Adds margins around the content of
/// a scrollable view. `edges` is parsed via the shared `RNWEdgeParser`
/// (defaults to `.all`); `length` is the inset in points. Applied for the
/// `.scrollContent` placement so it insets scrollable content specifically.
///
/// Gated to watchOS 10+. Returns the view unchanged on older systems.
enum RNWContentMarginsModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("contentMargins") { view, params, _ in
            if #available(watchOS 10.0, *) {
                let edges = RNWEdgeParser.edgeSet(params["edges"])
                let length = params.cgFloat("length") ?? 0
                return AnyView(view.contentMargins(edges, length, for: .scrollContent))
            }
            return view
        }
    }
}
